morpheus/pipeline/pipeline.py

```diff
@@ -215,6 +218,54 @@ def add_segment_edge(self,
             "output_receiver": ingress_stage.unique_name
         })
 
+    def pre_build(self):
+        assert not self.is_pre_built, "Pipeline can only be pre-built once!"
+        assert not self.is_built, "Pipeline cannot be built before pre-build"
+        assert len(self._sources) > 0, "Pipeline must have a source stage"
+
+        logger.info("====Pipeline Pre-build====")
+
+        for segment_id in self._segments.keys():
+            logger.info("====Pre-Building Segment: %s====", segment_id)
+            segment_graph = self._segment_graphs[segment_id]
+
+            # Check if preallocated columns are requested, this needs to happen before the source stages are built
+            needed_columns = OrderedDict()
+            for stage in networkx.topological_sort(segment_graph):
+                needed_columns.update(stage.get_needed_columns())
+
+            if (len(needed_columns) > 0):
+                for stage in segment_graph.nodes():
+                    if (isinstance(stage, PreallocatorMixin)):
+                        stage.set_needed_columns(needed_columns)
+
+            # This should be a BFS search from each source nodes; but, since we don't have source stage loops
+            # topo_sort provides a reasonable approximation.
+            for stage in networkx.topological_sort(segment_graph):
+                if (stage.can_pre_build()):
+                    stage._pre_build()
+
+            if (not all(x.is_built for x in segment_graph.nodes())):
+                logger.warning("Cyclic pipeline graph detected! Building with reduced constraints")
+
+                for stage in segment_graph.nodes():
+                    if (stage.can_pre_build(check_ports=True)):
+                        stage._pre_build()
+
+            if (not all(x.is_pre_built for x in segment_graph.nodes())):
+                raise RuntimeError("Could not build pipeline. Ensure all types can be determined")
```

Comment:
That I'm not sure of. This code existed in the original and is just being moved to a new method. 

In this case a pipeline failing to build like this would be extremely rare and I believe would require a cycle to be present, which I don't know we fully support yet.